index.html

```diff
@@ -548,73 +551,92 @@ <h2>
+            <li>Ensure that |options|'s {{PushSubscriptionOptionsInit/applicationServerKey}}
+            describes a valid point on the P-256 curve. If its value is invalid, return [=a promise
+            rejected with=] an {{"InvalidAccessError"}} {{DOMException}}.
             </li>
           </ol>
         </li>
-        <li>If the |options| argument does not include a non-null value for the
-        {{PushSubscriptionOptions/applicationServerKey}} attribute, and the <a>push service</a>
-        requires one to be given, reject |promise| with a {{DOMException}} whose name is
-        {{"NotSupportedError"}} and terminate these steps.
+        <li>Let |registration:ServiceWorkerRegistration| be the {{PushManager}}'s associated
+        <a>service worker registration</a>.
+        </li>
+        <li>If |registration|'s [=service worker registration/active worker=] is null, return [=a
+        promise rejected with=] an {{"InvalidStateError"}} {{DOMException}}.
+        </li>
+        <li>Let |sw| be |registration|'s [=service worker registration/active worker=].
+        </li>
+        <li>Let |promise| be [=a new promise=].
         </li>
-        <li>Let |registration| be the {{PushManager}}'s associated <a>service worker
-        registration</a>.
+        <li>Let |global:Window| be [=this=]'s [=relevant global object=].
         </li>
-        <li>If |registration|'s [=service worker registration/active worker=] is null, reject
-        |promise| with a {{DOMException}} whose name is {{"InvalidStateError"}} and terminate these
-        steps.
+        <li>Return |promise| and continue [=in parallel=].
         </li>
         <li>Let |permission| be [=request permission to use=] "push".
         </li>
-        <li>If |permission| is "denied", reject |promise| with a {{DOMException}} whose name is
-        {{"NotAllowedError"}} and terminate these steps.
+        <li>If |permission| is {{PermissionState/"denied"}}, [=queue a global task=] on the [=user
+        interaction task source=] using |global| to reject |promise| with a {{"NotAllowedError"}}
+        {{DOMException}} and terminate these steps.
         </li>
-        <li>If the <a>Service Worker</a> is already subscribed, run the following substeps:
+        <li>If |sw| is already subscribed, run the following substeps:
           <ol>
-            <li>Retrieve the <a>push subscription</a> associated with the <a>Service Worker</a>.
+            <li>Let |subscription| be the <a>push subscription</a> associated with |sw|.
             </li>
-            <li>If there is an error, reject |promise| with a {{DOMException}} whose name is
-            {{"AbortError"}} and terminate these steps.
+            <li>If |subscription| is an error, [=queue a global task=] on the [=networking task
+            source=] using |global| to [=reject=] |promise| with an {{"AbortError"}}
+            {{DOMException}} and terminate these steps.
```

Comment:
> How can subscription be an error here? 

You are right. It can't be an error. 

> The outcome of an attempt to retrieve it might return an error, but the subscription will be a subscription, right?

Right. In fact, so long as [=create a push subscription=] succeeded previously, there is _always_ a subscription. 

The problem (question?) is actually in [=create a push subscription=]: can [=create a push subscription=] result in an exception? 

It appears it can for a variety of reasons (e.g., [Gecko binding layer](https://searchfox.org/mozilla-central/source/__GENERATED__/dom/bindings/PushManagerBinding.cpp#1637) generates a bunch, [WebKit](https://webkit-search.igalia.com/webkit/source/Source/WebCore/workers/service/ServiceWorkerContainer.cpp#573)).

So, I think the right thing to do here is basically say, if creating [=create a push subscription=] results in an exception, reject the promise with it. 

I'll also update [=create a push subscription=] so it "tries" to create the push subscription and re-throws any exceptions.

Reply:
Sorry, to also clarify, you are right that the retrieval of the conceptual "subscription" might result in an error (basically, what happens with `getSubscription()`).